{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad84acda",
   "metadata": {},
   "source": [
    "This tutorial will demonstrate how to pre-process single-cell raw UMI counts to generate expression matrices that can be used as input to cell-cell communication tools. We will assume appropriate quality-control (QC) has already been applied to the dataset (e.g., exclusion of low-quality cells and doublets). We recommend the tutorial by [Luecken & Theis](https://doi.org/10.15252/msb.20188746) as a starting point for a detailed overview of QC and single-cell RNAseq analysis pipelines in general. \n",
    "\n",
    "Here we will focus on:\n",
    "1. Normalization\n",
    "2. Inter-operability between R and python. \n",
    "\n",
    "We demonstrate a typical workflow using the popular single-cell analysis software [scanpy](https://scanpy.readthedocs.io/en/stable/) to generate an AnnData object which can be used downstream. We will use a [BALF COVID dataset](https://doi.org/10.1038/s41591-020-0901-9), which contains 12 samples associated with \"Healthy Control\", \"Moderate\", or \"Severe\" COVID contexts.\n",
    "\n",
    "Details and caveats regarding [batch correction](https://www.nature.com/articles/s41592-018-0254-1), which removes technical variation while preserving biological variation between samples, can be viewed in the additional examples tutorial entitled \"S1_Batch_Correction\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "dbff2e24",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pickle\n",
    "\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import sys\n",
    "sys.path.insert(1, '/home/hratch/Projects/CCC/ccc_protocols/scripts/') # replace with cell2cell\n",
    "from cell2cell_dev.datasets.load_data import CovidBalf\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "seed = 888\n",
    "\n",
    "data_path = '/data3/hratch/ccc_protocols/'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2b704b19",
   "metadata": {},
   "source": [
    "The 12 samples can be downloaded as .h5 files from [here](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE145926). You can also download the cell metadata from [here](https://raw.githubusercontent.com/zhangzlab/covid_balf/master/all.cell.annotation.meta.txt)\n",
    "\n",
    "cell2cell has a helper function to download and format the raw UMI counts and metadata:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "564ab7b6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample_ID</th>\n",
       "      <th>sample_new</th>\n",
       "      <th>Context</th>\n",
       "      <th>disease</th>\n",
       "      <th>hasnCoV</th>\n",
       "      <th>cluster</th>\n",
       "      <th>cell_type</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCACAGCTACAT_3</th>\n",
       "      <td>C100</td>\n",
       "      <td>HC3</td>\n",
       "      <td>Healthy_Control</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>27</td>\n",
       "      <td>B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCATCCACGGGT_3</th>\n",
       "      <td>C100</td>\n",
       "      <td>HC3</td>\n",
       "      <td>Healthy_Control</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>23</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCCATCCCATTCG_3</th>\n",
       "      <td>C100</td>\n",
       "      <td>HC3</td>\n",
       "      <td>Healthy_Control</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>6</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAACAAACAGGC_3</th>\n",
       "      <td>C100</td>\n",
       "      <td>HC3</td>\n",
       "      <td>Healthy_Control</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>10</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTCGCACAC_3</th>\n",
       "      <td>C100</td>\n",
       "      <td>HC3</td>\n",
       "      <td>Healthy_Control</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>10</td>\n",
       "      <td>Macrophages</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   Sample_ID sample_new          Context disease hasnCoV  \\\n",
       "cell_barcode                                                               \n",
       "AAACCCACAGCTACAT_3      C100        HC3  Healthy_Control       N       N   \n",
       "AAACCCATCCACGGGT_3      C100        HC3  Healthy_Control       N       N   \n",
       "AAACCCATCCCATTCG_3      C100        HC3  Healthy_Control       N       N   \n",
       "AAACGAACAAACAGGC_3      C100        HC3  Healthy_Control       N       N   \n",
       "AAACGAAGTCGCACAC_3      C100        HC3  Healthy_Control       N       N   \n",
       "\n",
       "                    cluster    cell_type  \n",
       "cell_barcode                              \n",
       "AAACCCACAGCTACAT_3       27            B  \n",
       "AAACCCATCCACGGGT_3       23  Macrophages  \n",
       "AAACCCATCCCATTCG_3        6            T  \n",
       "AAACGAACAAACAGGC_3       10  Macrophages  \n",
       "AAACGAAGTCGCACAC_3       10  Macrophages  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "covid_balf_data = CovidBalf(data_path = os.path.join(data_path, 'raw/covid_balf/'))\n",
    "# covid_balf_data.download_data()\n",
    "md, balf_samples = covid_balf_data.format_data()\n",
    "\n",
    "md.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7c283d",
   "metadata": {},
   "source": [
    "balf_samples is a dictionary with keys as each sample and values as an AnnData object storing the raw UMI counts for that sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bb14f30a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['C100', 'C51', 'C52', 'C141', 'C142', 'C144', 'C143', 'C145', 'C146', 'C148', 'C149', 'C152'])"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "balf_samples.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5d757698",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2566 × 16566\n",
       "    obs: 'Sample_ID', 'Context', 'cell_type'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "balf_samples['C100']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "32c99290",
   "metadata": {},
   "source": [
    "To normalize the raw UMI counts, we recommend log(1+CPM) normalization, as this maintains non-negative counts and is the input for many communication scoring functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "daa05c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "for adata in balf_samples.values():\n",
    "    adata.raw = adata # store the raw UMI counts (note this is actually typically used to store normalized counts)\n",
    "    sc.pp.normalize_total(adata, target_sum=1e6) # CPM normalize\n",
    "    sc.pp.log1p(adata) # logarithmize"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "33d4420d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2566 × 16566\n",
       "    obs: 'Sample_ID', 'Context', 'cell_type'\n",
       "    var: 'gene_ids', 'feature_types', 'genome', 'n_cells'\n",
       "    uns: 'log1p'"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "balf_samples['C100']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b6d04d06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCCACAGCTACAT-1</th>\n",
       "      <th>AAACCCATCCACGGGT-1</th>\n",
       "      <th>AAACCCATCCCATTCG-1</th>\n",
       "      <th>AAACGAACAAACAGGC-1</th>\n",
       "      <th>AAACGAAGTCGCACAC-1</th>\n",
       "      <th>AAACGAAGTCTATGAC-1</th>\n",
       "      <th>AAACGAAGTGTAGTGG-1</th>\n",
       "      <th>AAACGCTGTCACGTGC-1</th>\n",
       "      <th>AAACGCTGTTGGAGGT-1</th>\n",
       "      <th>AAAGAACTCTAGAACC-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGATCTCCCGAAAT-1</th>\n",
       "      <th>TTTGGAGCAATACAGA-1</th>\n",
       "      <th>TTTGGAGTCACCATAG-1</th>\n",
       "      <th>TTTGGAGTCTCACCCA-1</th>\n",
       "      <th>TTTGGTTAGATGGCGT-1</th>\n",
       "      <th>TTTGGTTGTACCCAGC-1</th>\n",
       "      <th>TTTGGTTGTTACTCAG-1</th>\n",
       "      <th>TTTGTTGAGCTAGAGC-1</th>\n",
       "      <th>TTTGTTGCAATGAAAC-1</th>\n",
       "      <th>TTTGTTGCAGAGGGTT-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.492637</td>\n",
       "      <td>4.373578</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.111976</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.413881</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.855852</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.001009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.127682</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.041668</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.305098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2566 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAACCCACAGCTACAT-1  AAACCCATCCACGGGT-1  AAACCCATCCCATTCG-1  \\\n",
       "A1BG                     0.0                 0.0                 0.0   \n",
       "A1BG-AS1                 0.0                 0.0                 0.0   \n",
       "A2M                      0.0                 0.0                 0.0   \n",
       "A2M-AS1                  0.0                 0.0                 0.0   \n",
       "A2ML1                    0.0                 0.0                 0.0   \n",
       "\n",
       "          AAACGAACAAACAGGC-1  AAACGAAGTCGCACAC-1  AAACGAAGTCTATGAC-1  \\\n",
       "A1BG                3.492637            4.373578                 0.0   \n",
       "A1BG-AS1            0.000000            0.000000                 0.0   \n",
       "A2M                 4.855852            0.000000                 0.0   \n",
       "A2M-AS1             0.000000            0.000000                 0.0   \n",
       "A2ML1               0.000000            0.000000                 0.0   \n",
       "\n",
       "          AAACGAAGTGTAGTGG-1  AAACGCTGTCACGTGC-1  AAACGCTGTTGGAGGT-1  \\\n",
       "A1BG                     0.0                 0.0            4.111976   \n",
       "A1BG-AS1                 0.0                 0.0            0.000000   \n",
       "A2M                      0.0                 0.0            0.000000   \n",
       "A2M-AS1                  0.0                 0.0            0.000000   \n",
       "A2ML1                    0.0                 0.0            0.000000   \n",
       "\n",
       "          AAAGAACTCTAGAACC-1  ...  TTTGATCTCCCGAAAT-1  TTTGGAGCAATACAGA-1  \\\n",
       "A1BG                0.000000  ...                 0.0                 0.0   \n",
       "A1BG-AS1            0.000000  ...                 0.0                 0.0   \n",
       "A2M                 6.001009  ...                 0.0                 0.0   \n",
       "A2M-AS1             0.000000  ...                 0.0                 0.0   \n",
       "A2ML1               0.000000  ...                 0.0                 0.0   \n",
       "\n",
       "          TTTGGAGTCACCATAG-1  TTTGGAGTCTCACCCA-1  TTTGGTTAGATGGCGT-1  \\\n",
       "A1BG                5.413881                 0.0                 0.0   \n",
       "A1BG-AS1            0.000000                 0.0                 0.0   \n",
       "A2M                 5.127682                 0.0                 0.0   \n",
       "A2M-AS1             0.000000                 0.0                 0.0   \n",
       "A2ML1               0.000000                 0.0                 0.0   \n",
       "\n",
       "          TTTGGTTGTACCCAGC-1  TTTGGTTGTTACTCAG-1  TTTGTTGAGCTAGAGC-1  \\\n",
       "A1BG                0.000000                 0.0                 0.0   \n",
       "A1BG-AS1            0.000000                 0.0                 0.0   \n",
       "A2M                 5.041668                 0.0                 0.0   \n",
       "A2M-AS1             0.000000                 0.0                 0.0   \n",
       "A2ML1               0.000000                 0.0                 0.0   \n",
       "\n",
       "          TTTGTTGCAATGAAAC-1  TTTGTTGCAGAGGGTT-1  \n",
       "A1BG                     0.0            0.000000  \n",
       "A1BG-AS1                 0.0            0.000000  \n",
       "A2M                      0.0            5.305098  \n",
       "A2M-AS1                  0.0            0.000000  \n",
       "A2ML1                    0.0            0.000000  \n",
       "\n",
       "[5 rows x 2566 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ordered_genes = sorted(balf_samples['C100'].to_df().T.index)\n",
    "balf_samples['C100'].to_df().T.loc[ordered_genes,:].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9d77cec",
   "metadata": {},
   "source": [
    "We can save this diction for future use in other scripts:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0c80bfae",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(data_path, 'interim/covid_balf_norm.pickle'), 'wb') as handle:\n",
    "    pickle.dump(balf_samples, handle, protocol=pickle.HIGHEST_PROTOCOL)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cd3f788",
   "metadata": {},
   "source": [
    "# Interoperability"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1d4cf94",
   "metadata": {},
   "source": [
    "## to R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e280eecf",
   "metadata": {},
   "source": [
    "For use in R, we can save each AnnData object, and then load it in R as a SeuratObject using [SeuratDisk](https://mojaveazure.github.io/seurat-disk/articles/convert-anndata.html). See the companion R tutorial for loading these saved files. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "bb0daae3",
   "metadata": {},
   "outputs": [],
   "source": [
    "for sample in balf_samples:\n",
    "    file_name = os.path.join(data_path, 'interim/covid_python_to_R/', sample + '.h5ad')\n",
    "    adata = balf_samples[sample]\n",
    "    adata.write_h5ad(file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5094103",
   "metadata": {},
   "source": [
    "## from R\n",
    "\n",
    "Here, we can load the expression matrices that were generated in the companion R script using Seurat and saved as h5ad files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "2f7cc9cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "balf_samples_R = dict()\n",
    "for sample in balf_samples:\n",
    "    file_name = os.path.join(data_path, 'interim/covid_R_to_python/', sample + '.h5ad')\n",
    "    balf_samples_R[sample] = sc.read_h5ad(file_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4be80ca",
   "metadata": {},
   "source": [
    "While this AnnData object is not completely identical to the one generated in this script, it stores all the same information. Raw UMI counts are stored in the .raw attribute, relevant metadata is available in .obsm, and the log(1+CPM) matrix is stored in .X. We can see that the expression matrix is the same:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "80299e75",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 2566 × 16566\n",
       "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'Sample.ID', 'Context', 'cell.type'\n",
       "    var: 'features'"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "balf_samples_R['C100']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "973a1fc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAACCCACAGCTACAT-1</th>\n",
       "      <th>AAACCCATCCACGGGT-1</th>\n",
       "      <th>AAACCCATCCCATTCG-1</th>\n",
       "      <th>AAACGAACAAACAGGC-1</th>\n",
       "      <th>AAACGAAGTCGCACAC-1</th>\n",
       "      <th>AAACGAAGTCTATGAC-1</th>\n",
       "      <th>AAACGAAGTGTAGTGG-1</th>\n",
       "      <th>AAACGCTGTCACGTGC-1</th>\n",
       "      <th>AAACGCTGTTGGAGGT-1</th>\n",
       "      <th>AAAGAACTCTAGAACC-1</th>\n",
       "      <th>...</th>\n",
       "      <th>TTTGATCTCCCGAAAT-1</th>\n",
       "      <th>TTTGGAGCAATACAGA-1</th>\n",
       "      <th>TTTGGAGTCACCATAG-1</th>\n",
       "      <th>TTTGGAGTCTCACCCA-1</th>\n",
       "      <th>TTTGGTTAGATGGCGT-1</th>\n",
       "      <th>TTTGGTTGTACCCAGC-1</th>\n",
       "      <th>TTTGGTTGTTACTCAG-1</th>\n",
       "      <th>TTTGTTGAGCTAGAGC-1</th>\n",
       "      <th>TTTGTTGCAATGAAAC-1</th>\n",
       "      <th>TTTGTTGCAGAGGGTT-1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.492637</td>\n",
       "      <td>4.373578</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.111976</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.413881</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1BG-AS1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.855851</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.001009</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.127682</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.041668</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.305098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M-AS1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2ML1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 2566 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          AAACCCACAGCTACAT-1  AAACCCATCCACGGGT-1  AAACCCATCCCATTCG-1  \\\n",
       "A1BG                     0.0                 0.0                 0.0   \n",
       "A1BG-AS1                 0.0                 0.0                 0.0   \n",
       "A2M                      0.0                 0.0                 0.0   \n",
       "A2M-AS1                  0.0                 0.0                 0.0   \n",
       "A2ML1                    0.0                 0.0                 0.0   \n",
       "\n",
       "          AAACGAACAAACAGGC-1  AAACGAAGTCGCACAC-1  AAACGAAGTCTATGAC-1  \\\n",
       "A1BG                3.492637            4.373578                 0.0   \n",
       "A1BG-AS1            0.000000            0.000000                 0.0   \n",
       "A2M                 4.855851            0.000000                 0.0   \n",
       "A2M-AS1             0.000000            0.000000                 0.0   \n",
       "A2ML1               0.000000            0.000000                 0.0   \n",
       "\n",
       "          AAACGAAGTGTAGTGG-1  AAACGCTGTCACGTGC-1  AAACGCTGTTGGAGGT-1  \\\n",
       "A1BG                     0.0                 0.0            4.111976   \n",
       "A1BG-AS1                 0.0                 0.0            0.000000   \n",
       "A2M                      0.0                 0.0            0.000000   \n",
       "A2M-AS1                  0.0                 0.0            0.000000   \n",
       "A2ML1                    0.0                 0.0            0.000000   \n",
       "\n",
       "          AAAGAACTCTAGAACC-1  ...  TTTGATCTCCCGAAAT-1  TTTGGAGCAATACAGA-1  \\\n",
       "A1BG                0.000000  ...                 0.0                 0.0   \n",
       "A1BG-AS1            0.000000  ...                 0.0                 0.0   \n",
       "A2M                 6.001009  ...                 0.0                 0.0   \n",
       "A2M-AS1             0.000000  ...                 0.0                 0.0   \n",
       "A2ML1               0.000000  ...                 0.0                 0.0   \n",
       "\n",
       "          TTTGGAGTCACCATAG-1  TTTGGAGTCTCACCCA-1  TTTGGTTAGATGGCGT-1  \\\n",
       "A1BG                5.413881                 0.0                 0.0   \n",
       "A1BG-AS1            0.000000                 0.0                 0.0   \n",
       "A2M                 5.127682                 0.0                 0.0   \n",
       "A2M-AS1             0.000000                 0.0                 0.0   \n",
       "A2ML1               0.000000                 0.0                 0.0   \n",
       "\n",
       "          TTTGGTTGTACCCAGC-1  TTTGGTTGTTACTCAG-1  TTTGTTGAGCTAGAGC-1  \\\n",
       "A1BG                0.000000                 0.0                 0.0   \n",
       "A1BG-AS1            0.000000                 0.0                 0.0   \n",
       "A2M                 5.041668                 0.0                 0.0   \n",
       "A2M-AS1             0.000000                 0.0                 0.0   \n",
       "A2ML1               0.000000                 0.0                 0.0   \n",
       "\n",
       "          TTTGTTGCAATGAAAC-1  TTTGTTGCAGAGGGTT-1  \n",
       "A1BG                     0.0            0.000000  \n",
       "A1BG-AS1                 0.0            0.000000  \n",
       "A2M                      0.0            5.305098  \n",
       "A2M-AS1                  0.0            0.000000  \n",
       "A2ML1                    0.0            0.000000  \n",
       "\n",
       "[5 rows x 2566 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ordered_genes = sorted(balf_samples_R['C100'].to_df().T.index)\n",
    "balf_samples_R['C100'].to_df().T.loc[ordered_genes,:].head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ccc_protocols]",
   "language": "python",
   "name": "conda-env-ccc_protocols-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
